fs/fssnapshotfs/snapshot_fs.go

```diff
@@ -0,0 +1,305 @@
+package fssnapshotfs
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	fs_snapshot "github.com/pescuma/go-fs-snapshot/lib"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/kopia/kopia/fs/localfs"
+	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/snapshot/policy"
+)
+
+type snapshotEntry struct {
+	policyTree      *policy.Tree
+	snapshoter      fs_snapshot.Snapshoter
+	backuper        fs_snapshot.Backuper
+	closeSnapshoter bool
+	originalRoot    string
+	snapshotRoot    string
+	fs.Entry
+}
+
+// NewEntry returns fs.Entry for the specified path, the result will be one of supported entry types: fs.File, fs.Directory, fs.Symlink
+// or fs.UnsupportedEntry.
+func NewEntry(originalPath string, policyTree *policy.Tree) (fs.Entry, error) {
+	entry, err := localfs.NewEntry(originalPath)
+	if err != nil {
+		//nolint:wrapcheck
+		return nil, err
+	}
+
+	cfg := &fs_snapshot.SnapshoterConfig{
+		InfoCallback: func(level fs_snapshot.MessageLevel, msg string, a ...interface{}) {
+			if level == fs_snapshot.OutputLevel {
+				fmt.Printf(msg+"\n", a...)
+			}
+		},
+	}
+
+	snapshoter, err := fs_snapshot.NewSnapshoter(cfg)
+	if err != nil {
+		// If we can't use snapshots, just use the filesystem directly
+		fmt.Printf("Error creating filesystem snapshot: %v\n", err)
+
+		return entry, nil
+	}
+
+	backuper, err := snapshoter.StartBackup(nil)
+	if err != nil {
+		// If we can't use snapshots, just use the filesystem directly
```

Comment:
I think we should actually fail hard if snapshot is requested but could not be made.

Reply:
I like the idea of just leaving it enabled and it works where available. Maybe make it have 3 options: required, if available, no ?

Reply:
sure, that's gonna work, as long as we don't silently fall back to inconsistent snapshots when the user thinks we're snapshotting.

Reply:
BTW, when we fail to create FS snapshot and decide to move forward with it, this should be recorded somewhere.

In fact, we should probably record details of snapshot on `*snapshot.Manifest` in all cases, something like:

```
"rootEntry": {
   ...
},
"filesystemSnapshot": {
  "id": "..."
  "timestamp": "...",
  "method": "VSS",
}
```

or

```
"rootEntry": {
   ...
},
"filesystemSnapshot": {
  "method": "VSS",
  "error": {
    "message": "Unable to create VSS snapshot because ..."
   }
}
```

Reply:
I added logging when creating/an error occurred. But I don't have the snapshot manifest at hand to add more info to it. Is there an easy way to do that? Another thing is that it can contain multiple snapshots (when there are different drives mounted inside the snapshot root), so this have to be a list.

Reply:
Ok, after the change in approach it became easy to do. Now it adds info to the manifest too.